In [19]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 1: Report Repair ---</h2><p>After saving Christmas <a href="/events">five years in a row</a>, you've decided to take a vacation at a nice resort on a tropical island. <span title="WHAT COULD GO WRONG">Surely</span>, Christmas will go on without you.</p>
<p>The tropical island has its own currency and is entirely cash-only.  The gold coins used there have a little picture of a starfish; the locals just call them <em class="star">stars</em>. None of the currency exchanges seem to have heard of them, but somehow, you'll need to find fifty of these coins by the time you arrive so you can pay the deposit on your room.</p>
<p>To save your vacation, you need to get all <em class="star">fifty stars</em> by December 25th.</p>
<p>Collect stars by solving puzzles.  Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. Good luck!</p>
<p>Before you leave, the Elves in accounting just need you to fix your <em>expense report</em> (your puzzle input); apparently, something isn't quite adding up.</p>
<p>Specifically, they need you to <em>find the two entries that sum to <code>2020</code></em> and then multiply those two numbers together.</p>
<p>For example, suppose your expense report contained the following:</p>
<pre><code>1721
979
366
299
675
1456
</code></pre>
<p>In this list, the two entries that sum to <code>2020</code> are <code>1721</code> and <code>299</code>. Multiplying them together produces <code>1721 * 299 = 514579</code>, so the correct answer is <code><em>514579</em></code>.</p>
<p>Of course, your expense report is much larger. <em>Find the two entries that sum to <code>2020</code>; what do you get if you multiply them together?</em></p>
</article>


In [20]:
from sys import prefix


example = """
    1721
    979
    366
    299
    675
    1456
"""


def product_of_two_entries_with_sum_2020(data: str) -> int:
    ints = [int(l.strip()) for l in data.strip().splitlines()]
    prefix_sums = set()
    for i in ints:
        if 2020 - i in prefix_sums:
            return i * (2020 - i)
        prefix_sums.add(i)


assert product_of_two_entries_with_sum_2020(example) == 514579

In [21]:
with open("../input/day1.txt") as f:
    puzzle = f.read()

print(f"Part I: {product_of_two_entries_with_sum_2020(puzzle)}")

Part I: 974304


<link href="style.css" rel="stylesheet"></link>
<main class="read-aloud">

<p>Your puzzle answer was <code>974304</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The Elves in accounting are thankful for your help; one of them even offers you a starfish coin they had left over from a past vacation. They offer you a second one if you can find <em>three</em> numbers in your expense report that meet the same criteria.</p>
<p>Using the above example again, the three entries that sum to <code>2020</code> are <code>979</code>, <code>366</code>, and <code>675</code>. Multiplying them together produces the answer, <code><em>241861950</em></code>.</p>
<p>In your expense report, <em>what is the product of the three entries that sum to <code>2020</code>?</em></p>
</article>

</main>


In [22]:
from math import prod
from more_itertools import one


def product_of_three_entries_with_sum_2020(data: str) -> int:
    nums = [int(l.strip()) for l in data.strip().splitlines()]
    triplets = []
    nums.sort()
    for i in range(len(nums)):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        sum = 2020 - nums[i]
        j = i + 1
        k = len(nums) - 1
        while j < k:
            sum2 = nums[j] + nums[k]

            if sum2 == sum:
                triplets.append(sorted([nums[i], nums[j], nums[k]]))
                while j < k and nums[j] == nums[j + 1]:
                    j += 1
                while j < k and nums[k - 1] == nums[k]:
                    k -= 1
                j += 1
                k -= 1
            elif sum2 < sum:
                j += 1
            else:
                k -= 1

    return prod(one(triplets))


assert product_of_three_entries_with_sum_2020(example) == 241861950

In [23]:
print(f"Part II: {product_of_three_entries_with_sum_2020(puzzle)}")

Part II: 236430480


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>236430480</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
